pkg/streamingpromql/types/fpoint_ring_buffer.go

```diff
@@ -294,6 +304,124 @@ func (v FPointRingBufferView) Clone() (*FPointRingBufferView, *FPointRingBuffer,
 	return view, buffer, nil
 }
 
+// FPointRingBufferViewIterator is an iterator which can be used over a FPointRingBufferView
+type FPointRingBufferViewIterator struct {
+	idx  int
+	view *FPointRingBufferView
+}
+
+func NewFPointRingBufferViewIterator(view *FPointRingBufferView) *FPointRingBufferViewIterator {
+	return &FPointRingBufferViewIterator{view: view}
+}
+
+func (i *FPointRingBufferViewIterator) HasNext() bool {
+	return i.idx < i.view.Count()
+}
+
+// Next moves the iterator forward, returning the next point.
+// This function will panic if moving the iterator would result in an index out of bounds.
+func (i *FPointRingBufferViewIterator) Next() promql.FPoint {
+	if i.idx >= i.view.Count() {
+		panic(fmt.Sprintf("next(): out of range, requested index %v but have length %v", i.idx, i.view.Count()))
+	}
+	p := i.view.PointAt(i.idx)
+	i.idx++
+	return p
+}
+
+func (i *FPointRingBufferViewIterator) At() promql.FPoint {
+	return i.view.PointAt(i.idx)
+}
+
+// Prev moves the iterator backwards, returning the previous point.
+// This function will panic if moving the iterator would result in an index out of bounds.
+func (i *FPointRingBufferViewIterator) Prev() promql.FPoint {
+	if i.idx <= 0 {
+		panic(fmt.Sprintf("prev(): out of range, requested index %v", i.idx-1))
+	}
+	i.idx--
+	return i.view.PointAt(i.idx)
+}
+
+// advance will move the iterator forward.
+func (i *FPointRingBufferViewIterator) advance() {
+	if i.idx < i.view.Count() {
+		i.idx++
+	}
+}
+
+// reverse will move the iterator backwards.
+func (i *FPointRingBufferViewIterator) reverse() {
+	if i.idx > 0 {
+		i.idx--
+	}
+}
+
+// Peek returns the next point, but does not move the iterator forward.
+// This function will panic if this look ahead would result in an index out of bounds.
+func (i *FPointRingBufferViewIterator) Peek() promql.FPoint {
+	if i.idx >= i.view.Count() {
+		panic(fmt.Sprintf("peek(): out of range, requested index %v but have length %v", i.idx, i.view.Count()))
+	}
+	return i.view.PointAt(i.idx)
+}
+
+// Seek will return the point which is closest to being <= time, or the first point after this time.
+// The iterator will be positioned to the first point > time.
+func (i *FPointRingBufferViewIterator) Seek(time int64) promql.FPoint {
+	var first *promql.FPoint
+
+	for i.HasNext() {
+		next := i.Peek()
+
+		if next.T < time {
+			first = &next
```

Comment:
I suspect this is causing a lot of garbage, which might be the cause of the performance issues you're seeing - what happens if you change `first` to be a value (ie. not a pointer) and use a `bool` to track if we've seen any points?